src/v/cloud_storage/async_manifest_view.cc

```diff
@@ -896,12 +941,32 @@ async_manifest_view::size_based_retention(size_t size_limit) noexcept {
                 if (!eof) {
                     auto r = co_await cursor->next();
                     if (
```

Comment:
Maybe it'd be clearer if we changed allowed the while loop to check the STM manifest, but explicitly handle the STM manifest case? Rather than having to reason about what next()==false means

Reply:
It used to be that the outer loop allowed the cursor to reach the stm manifest. Evgeny changed it at some point due to concerns that we could incorrectly apply retention within the stm region here. I think that's valid, so I was trying to work with it.